### ISSUE TITLE
LaTeX output rendering in sagemath only working when unblocking cloudflare.com and npmjs.org in Firefox's NoScript

### ISSUE THREAD

**Issue by chwrtlprmft** [jupyterlab/jupyterlab#9911]
## Description

I have to unblock _cloudflare.com_ and _npmjs.org_ (in addition to _localhost_) in NoScript extension for the Firefox tab JupyterLab runs in in order to make LaTeX formatted output work in my sagemath notebooks. This is _not_ the case in Jupyter Notebook, from where I just transitioned. If I just unblock _localhost_ in JupyterLab, there is just no output at all when specifying `%display latex`:

![screenshot_2021-03-07-102652](https://user-images.githubusercontent.com/72135724/110235469-62f03800-7f30-11eb-9d3a-e83df8a0265f.png)

Is that expected behaviour? Is LaTeX rendering done online in JupyterLab? If so, what are the privacy implications?

## Reproduce

1. On Arch Linux, install via `pacman -S --noconfirm sagemath sagemath-doc sagemath-jupyter jupyterlab`
2. In a terminal, run `sage -n jupyterlab`
3. Open one of the provided links in Firefox
4. Block cloudflare.com and npmjs.org using NoScript extension of Firefox (this is default)
5. Request LaTeX output using `%display latex`

## Expected behavior

Pi is rendered as LaTeX and no remote scripts from cloudflare.com and npmjs.org are needed for using jupyterlab.

## Context

- Operating System and version: Arch Linux 5.11.2-arch1-1
- Browser and version: Firefox 86.0
- JupyterLab version: Jupyter Server 1.4.1, JupyterLab Version 3.0.9

<details><summary>Troubleshoot Output</summary>
<pre>
Paste the output from running `jupyter troubleshoot` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.
</pre>
</details>

<details><summary>Browser Output</summary>
<pre>
Paste the output from your browser Javascript console here.
</pre>
</details>

**Comment by agoose77** [comment]
I am not sure where sagemath implements the LaTeX rendering, but I suspect what's happening here is that it doesn't use the Jupyter-provided endpoint to serve MathJax, and is instead making a direct call to these CDNs. In other words, this is probably a COCALC thing.

**Comment by agoose77** [comment]
Thanks, I've updated my comment.

**Comment by chwrtlprmft** [comment]
Thanks for the hints so far!

Yes, it happens too, if I run `$ jupyter lab` instead of `$ sage -n jupyterlab` to start my JupyterLab server and it happens too, if I request LaTeX output when working with the plain Python3 kernel in the JupyterLab session started with `$ jupyter lab`. No LaTeX is rendered when blocking cloudflare and npmjs.

Here what Firefox shows in the Network Tab of the Web Tools in case of $ sage -n jupyterlab` when working with the sage kernel:
![screenshot_2021-03-09-180820](https://user-images.githubusercontent.com/72135724/110511723-ce8afe80-8104-11eb-91ef-b060ef04100f.png)
![screenshot_2021-03-09-180838](https://user-images.githubusercontent.com/72135724/110511752-d8acfd00-8104-11eb-9d83-337caa3fc961.png)

This is the corresponding result for the plain Python3 session with `$ jupyter lab`:
![screenshot_2021-03-09-181637](https://user-images.githubusercontent.com/72135724/110511861-f9755280-8104-11eb-8d0c-5986814eae7c.png)

**Comment by jasongrout** [comment]
The npmjs url is a request from the extension manager to find jlab extensions you could install. You can turn off the extension manager in Settings, or disable it from its sidebar tab.

As for the cloudflare url, we right now fall back to CDN if the server does not have a local mathjax url configured: https://github.com/jupyterlab/jupyterlab_server/blob/a82794a1f4a80ed44f5385a16cb364eb452d1ad0/jupyterlab_server/handlers.py#L87-L92

This mathjax_url comes from the nbclassic package, which JupyterLab should be relying on: https://github.com/jupyterlab/nbclassic/blob/46bb6d55f48657331e71fc60f283b2fb38ab5557/nbclassic/traits.py#L113-L118. However, I see that you are installing jupyterlab not through pip, but through pacman. Perhaps it is not installing nbclassic, and that is what is going on here?

I think you can install the mathjax files into the local jupyter server directly (not requiring nbclassic) with https://github.com/jupyter-server/jupyter_server_mathjax

**Comment by chwrtlprmft** [comment]
Thanks for the detailed explanation. Arch installs `jupyter-nbclassic` as a dependency of `jupyterlab` by default via `pacman`. [Here](https://archlinux.org/packages/community/any/jupyter-nbclassic/files/) is the file list of that package.

I tried installing `jupyter_server_mathjax` on top using pip:
```
$ pacman -Syu --noconfirm
$ pacman -S --noconfirm python-pip sagemath sagemath-jupyter jupyterlab
$ yes | pip install jupyter_server_mathjax
```
But still no LaTeX rendering without unblocking cloudflare (i.e. no offline LaTeX rendering either).

For the npmjs request I went to the Extension Manager section n JupyterLab's advanced settings and set the user preferences to
```
{
    "disclaimed": false,
    "enabled": true
}
```

However `npmjs.org` still appears in NoScript after restarting JupyterLab...

**Comment by agoose77** [comment]
@chwrtlprmft can you share a reproducer notebook, and list the extensions in the extension manager?

**Comment by chwrtlprmft** [comment]
The extension manager is disabled (i enabled it to check, no installed extensions are listed). I'll attach a notebook below.

Here's a notebook to reproduce (or find it in the attached ZIP):
```
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/latex": [
       "$x=1$\n"
      ],
      "text/plain": [
       "<IPython.core.display.Latex object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%latex\n",
    "$x=1$"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
```

Here's also a recipe to build a Singularity container in case anyone wants to check what exactly is delivered with Arch in a quickly disposable environment:
```
BootStrap: docker
From: archlinux:latest

%post
	pacman -Syu --noconfirm
	pacman -S --noconfirm python-pip sagemath sagemath-jupyter jupyterlab
	yes | pip install jupyter_server_mathjax

%runscript
	sage -n jupyterlab

```

[reproducer.zip](https://github.com/jupyterlab/jupyterlab/files/6117309/reproducer.zip)

**Comment by agoose77** [comment]
I haven't the time to look at this just yet, but here's a Dockerfile for anyone using Docker/buildah

```Dockerfile
FROM archlinux:latest

RUN pacman -Syu --noconfirm
RUN pacman -S --noconfirm python-pip sagemath sagemath-jupyter jupyterlab
RUN yes | pip install jupyter_server_mathjax
ENTRYPOINT ["sage", "-n", "jupyterlab"]
```

**Comment by xavierq-pro** [comment]
Not sure if it is related but I add my two cents just in case:

System: OS = Void Linux (rolling release) ; JupyterLab = 3.0.10/Jupyter Server = 1.41/Jupyter Notebook = 6.2.0 ; browser = Firefox 86 - same one which might be important

Same error with MathJax not working OFFLINE in JupyterLab - but everything fine with Jupyter Notebook.

After some research:
- in Firefox web console, error messages:
       * Some cookies are misusing the recommended “SameSite“ attribute
       * Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://registry.npmjs.org/-/v1/search?text=+keywords%3A%22jupyterlab-extension%22&size=250&from=0. (Reason: CORS request did not succeed)

NB: other message differs depending whether or not I comment out what concerns cdnjs in file mentionned in previous post, it will say:
      * either "@jupyterlab/mathjax2-extension : plugin failed to activate" if commented out
      * or "Loading failed for the <script> with source https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-AMS_HTML-full,Safe&amp;delayStartupUntil=configured”

- following  https://jupyterlab.readthedocs.io/en/stable/getting_started/issue.html to troubleshoot:
       * step 3: unable to lauch Classic Notebook from JupyterLab (!) - 404: Not Found You are requesting a page that does not exist !
       * step 4: able to reproduce the problem in Firefox private browsing mode

- I tried to downgrade to Firefox 85, to no avail

- I installed jupyter_server_mathjax with no change to the result - normal if my assessment is correct

After googling "firefox cors request did not succeed" & "firefox misusing samesite attribute", I would say that a problem with the cookie policy (samesite attribute) somehow breaks the link between the different components of Jupyter (Lab/Server/Notebook).

Might be wrong (hopefully - maybe a problem with my settings ?). Any help appreciated.

**Comment by jasongrout** [comment]
I fixed the spurious npm registry request in https://github.com/jupyterlab/jupyterlab/pull/9974 (before a request was made by default even if the disclaimer was not agreed to, but now it is only made if the disclaimer is agreed to).

**Comment by jasongrout** [comment]
Tracking down the mathjax url issue will need some debugging:

1. First, to double-check the url that jlab is getting, can you run this in your browser js console?
   ```js
    JSON.parse(document.getElementById('jupyter-config-data').innerText).fullMathjaxUrl
    ```
    For a default jlab install, that should be something like `/static/notebook/components/MathJax/MathJax.js`. If it is pointing to cloudflare, which seems likely in your case, then that still points to a server misconfiguration somewhere.
2. Just to make sure this isn't a problem with how Sage or arch is packaging jlab, can you `pip install` jlab from the original python jlab package in a fresh environment and check the behavior there?

**Comment by chwrtlprmft** [comment]
Thanks for fixing the npm request!

2. I builld a new environment using this recipe:
```
BootStrap: docker
From: archlinux:latest

%post
	pacman -Syu --noconfirm
	pacman -S --noconfirm python-pip

	yes | pip install jupyter

```

Running Jupyter server from this container, I get not TeX rendering in Python3 notebooks (did not bother installing sage yet), the TeX code is just returned unchanged from cells with `%%latex` magic.

1. This is returned from the js command you provided: `"/static/components/MathJax/MathJax.js"`

**Comment by jasongrout** [comment]
Does math rendering work in markdown cells? Like a markdown cell with contents `$$\sqrt{x^2}$$` ?

**Comment by jasongrout** [comment]
> This is returned from the js command you provided: `"/static/components/MathJax/MathJax.js"`

That looks right. Is this installation requesting mathjax things from cloudflare? It shouldn't be.

**Comment by chwrtlprmft** [comment]
Sorry, I made a mistake and did not enclose the TeX code with `$$ $$` in the code cells (i normally just use sage in jlab an issue `display latex` once in the beginning):

Math rendering works finde from both markdown and code cells on that install without any call to cloudflare.

I think I should open an issue over at where the Arch package is tracked. Thanks a lot

**Comment by xavierq-pro** [comment]
Sorry to interfere again.

Did you test with Jupyter Server only ? As far as I can tell the problem is with Jupyter Lab, not Server/Notebook.

**Comment by chwrtlprmft** [comment]
No, I tested with jupyterlab, not jupyter notebook, i.e. `$ jupyter-lab` on the command line and then opened one of the issued links in Firefox.

**Comment by antonio-rojas** [comment]
Hi, Arch packager here. I'm seeing the exact same behavior in our packages and in a pip-installed jupyterlab: mathjax is loaded from cloudflare instead of using the local copy. The output of
`JSON.parse(document.getElementById('jupyter-config-data').innerText).fullMathjaxUrl`
in both cases is
`"https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js"`

**Comment by antonio-rojas** [comment]
Installing jupyter-server-mathjax works, in both pip and distro packages. So it seems the nbclassic codepath is broken.

**Comment by jasongrout** [comment]
Can you give us some more info to reproduce the problem? I'm not seeing it on pip-installed jupyterlab. What versions of the relevant packages, etc.?

**Comment by jasongrout** [comment]
For example, when I do a fresh `pip install jupyterlab` in a clean environment with python 3.9.2, the following packages are installed and I see `SON.parse(document.getElementById('jupyter-config-data').innerText).fullMathjaxUrl` give `"/static/notebook/components/MathJax/MathJax.js"`:

```
jupyter-client      6.1.12
jupyter-core        4.7.1
jupyter-packaging   0.7.12
jupyter-server      1.5.1
jupyterlab          3.0.12
jupyterlab-pygments 0.1.2
jupyterlab-server   2.3.0
...
nbclassic           0.2.6
nbclient            0.5.3
nbconvert           6.0.7
nbformat            5.1.2
nest-asyncio        1.5.1
notebook            6.3.0
```

**Comment by antonio-rojas** [comment]
```
> pip list --user
Package             Version
------------------- -------
anyio               2.2.0
argon2-cffi         20.1.0
async-generator     1.10
bleach              3.3.0
defusedxml          0.7.1
ipykernel           5.5.0
json5               0.9.5
jsonschema          3.2.0
jupyter-client      6.1.12
jupyter-core        4.7.1
jupyter-packaging   0.7.12
jupyter-server      1.5.1
jupyterlab          3.0.12
jupyterlab-pygments 0.1.2
jupyterlab-server   2.3.0
mistune             0.8.4
nbclassic           0.2.6
nbclient            0.5.3
nbconvert           6.0.7
nbformat            5.1.2
nest-asyncio        1.5.1
notebook            6.3.0
pandocfilters       1.4.3
prometheus-client   0.9.0
pyrsistent          0.17.3
pyzmq               22.0.3
Send2Trash          1.5.0
sniffio             1.2.0
terminado           0.9.3
testpath            0.4.4
tornado             6.1
```

**Comment by jasongrout** [comment]
> `> pip list --user`

Ah! There's the difference: `--user`. When I create a new conda python environment and use `pip install --user jupyterlab`, I do indeed see the mathjax coming from cloudflare.

Running `jupyter notebook` from the same `pip install --user` installation, I see it is getting its mathjax from localhost.

**Comment by jasongrout** [comment]
@xavierq-pro - that is a different issue than the one being discussed here. Can you open a new issue for it?

**Comment by xavierq-pro** [comment]
Yes, sorry but it shouldn't be necessary:

I had an old path set up in my jupyter_notebook_config.py (`/static/components/MathJax/MathJax.js`) which was interfering with the new (good ?) one (`/static/notebook/components/MathJax/MathJax.js`)

By removing it (the old path in config file), It looks like everything works for me - no more  error described above -  but this is weird because Mathjax is in /static/components... (not in /static/notebook/components which doesn't even exist)
